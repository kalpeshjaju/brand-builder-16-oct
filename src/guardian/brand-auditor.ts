// Brand Strategy Auditor - Complete quality validation

import type { BrandStrategy, AuditResult, AuditOptions } from '../types/index.js';
import { FactExtractor } from './fact-extractor.js';
import { SourceQualityAssessor } from './source-quality-assessor.js';
import { Logger } from '../utils/index.js';

const logger = new Logger('BrandAuditor');

export class BrandAuditor {
  private factExtractor: FactExtractor;
  private sourceAssessor: SourceQualityAssessor;

  constructor() {
    this.factExtractor = new FactExtractor();
    this.sourceAssessor = new SourceQualityAssessor();
  }

  /**
   * Audit a brand strategy
   */
  async audit(
    strategy: BrandStrategy,
    brandName: string,
    options: AuditOptions = {}
  ): Promise<AuditResult> {
    logger.info('Starting brand audit', { brandName, mode: options.mode });

    const startTime = Date.now();

    // Extract all text from strategy
    const strategyText = JSON.stringify(strategy);

    // Extract facts
    const facts = this.factExtractor.extractFacts(strategyText);
    const highConfidenceFacts = this.factExtractor.getHighConfidenceFacts(facts);

    // Assess sources
    const sources = (strategy.proofPoints || []).map((pp) => ({
      url: pp.sourceUrl || '',
      title: pp.source || '',
    })).filter((s) => s.url);

    const assessedSources = this.sourceAssessor.assessMultipleSources(sources);
    const avgTier = this.sourceAssessor.getAverageTier(assessedSources);

    // Calculate scores
    const sourceQualityScore = this.calculateSourceQualityScore(assessedSources.length, avgTier);
    const factVerificationScore = this.calculateFactVerificationScore(facts.length, highConfidenceFacts.length);
    const productionReadinessScore = this.calculateProductionReadinessScore(strategy);

    // Overall score (weighted average)
    const overallScore =
      sourceQualityScore * 0.3 +
      factVerificationScore * 0.25 +
      productionReadinessScore * 0.45;

    // Generate findings
    const findings = [
      ...this.generateSourceFindings(assessedSources),
      ...this.generateFactFindings(facts, highConfidenceFacts),
      ...this.generateProductionFindings(strategy),
    ];

    // Generate recommendations
    const recommendations = this.generateRecommendations(
      sourceQualityScore,
      factVerificationScore,
      productionReadinessScore
    );

    const duration = Date.now() - startTime;
    logger.info('Audit complete', { duration: `${duration}ms`, score: overallScore.toFixed(1) });

    return {
      brandName,
      auditDate: new Date().toISOString(),
      overallScore: Math.round(overallScore * 10) / 10,
      scoreBreakdown: {
        sourceQuality: {
          score: sourceQualityScore,
          weight: 0.3,
          status: this.getStatus(sourceQualityScore),
          details: `${assessedSources.length} sources assessed, avg tier: ${avgTier.toFixed(1)}`,
        },
        factVerification: {
          score: factVerificationScore,
          weight: 0.25,
          status: this.getStatus(factVerificationScore),
          details: `${highConfidenceFacts.length}/${facts.length} high-confidence facts`,
        },
        dataRecency: {
          score: 7,
          weight: 0.15,
          status: 'good',
          details: 'Data freshness assessment',
        },
        crossVerification: {
          score: 6,
          weight: 0.15,
          status: 'good',
          details: 'Cross-source verification',
        },
        productionReadiness: {
          score: productionReadinessScore,
          weight: 0.15,
          status: this.getStatus(productionReadinessScore),
          details: 'Strategy completeness check',
        },
      },
      findings,
      recommendations,
      qualityImprovement: {
        currentScore: Math.round(overallScore * 10) / 10,
        targetScore: 9.0,
        steps: recommendations.map((r, i) => ({
          step: i + 1,
          action: r.action,
          expectedImprovement: 0.5,
          estimatedTime: r.estimatedEffort,
        })),
        totalEffort: '8-12 hours',
        requiredExpertise: 'Mid-level analyst',
      },
    };
  }

  private calculateSourceQualityScore(count: number, avgTier: number): number {
    if (count === 0) return 2;
    const tierScore = (5 - avgTier) * 2.5; // Convert tier to 0-10 scale
    const countBonus = Math.min(count / 5, 1) * 2; // Bonus for having multiple sources
    return Math.min(tierScore + countBonus, 10);
  }

  private calculateFactVerificationScore(total: number, highConf: number): number {
    if (total === 0) return 5;
    const ratio = highConf / total;
    return ratio * 10;
  }

  private calculateProductionReadinessScore(strategy: BrandStrategy): number {
    let score = 0;
    if (strategy.purpose) score += 2;
    if (strategy.mission) score += 2;
    if (strategy.values && strategy.values.length > 0) score += 2;
    if (strategy.positioning) score += 2;
    if (strategy.keyMessages && strategy.keyMessages.length > 0) score += 2;
    return score;
  }

  private getStatus(score: number): 'excellent' | 'good' | 'needs-work' | 'critical' {
    if (score >= 8) return 'excellent';
    if (score >= 6) return 'good';
    if (score >= 4) return 'needs-work';
    return 'critical';
  }

  private generateSourceFindings(sources: any[]) {
    const findings = [];
    if (sources.length === 0) {
      findings.push({
        severity: 'warning' as const,
        category: 'sources' as const,
        message: 'No sources provided - add citations for credibility',
      });
    } else if (sources.length < 3) {
      findings.push({
        severity: 'info' as const,
        category: 'sources' as const,
        message: `Only ${sources.length} sources - consider adding more for cross-verification`,
      });
    }
    return findings;
  }

  private generateFactFindings(facts: any[], highConf: any[]) {
    const findings = [];
    if (facts.length > 0 && highConf.length < facts.length * 0.7) {
      findings.push({
        severity: 'warning' as const,
        category: 'facts' as const,
        message: 'Some extracted facts have low confidence - verify claims',
      });
    }
    return findings;
  }

  private generateProductionFindings(strategy: BrandStrategy) {
    const findings = [];
    if (!strategy.purpose) {
      findings.push({
        severity: 'critical' as const,
        category: 'quality' as const,
        message: 'Missing brand purpose - critical for strategy foundation',
      });
    }
    if (!strategy.positioning) {
      findings.push({
        severity: 'critical' as const,
        category: 'quality' as const,
        message: 'Missing positioning statement - define competitive position',
      });
    }
    return findings;
  }

  private generateRecommendations(source: number, fact: number, prod: number) {
    const recommendations = [];

    if (source < 7) {
      recommendations.push({
        priority: 'high' as const,
        action: 'Add high-quality sources (tier 1-2) for key claims',
        estimatedEffort: '2-4 hours',
        impact: 'Significantly increases credibility and trust',
      });
    }

    if (fact < 7) {
      recommendations.push({
        priority: 'medium' as const,
        action: 'Verify and strengthen fact-based claims with data',
        estimatedEffort: '3-5 hours',
        impact: 'Improves accuracy and defensibility',
      });
    }

    if (prod < 8) {
      recommendations.push({
        priority: 'high' as const,
        action: 'Complete missing strategy components (purpose, positioning, etc.)',
        estimatedEffort: '4-6 hours',
        impact: 'Essential for production-ready strategy',
      });
    }

    return recommendations;
  }
}
