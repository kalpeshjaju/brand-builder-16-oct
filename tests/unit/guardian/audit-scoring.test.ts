// Audit Scoring Snapshot Tests
// Tests that validate audit score calculations are deterministic and reproducible

import { describe, it, expect } from 'vitest';
import { BrandAuditor } from '../../../src/guardian/brand-auditor.js';
import type { BrandStrategy } from '../../../src/types/index.js';

describe('Audit Scoring Arithmetic', () => {
  const auditor = new BrandAuditor();

  describe('Score calculation determinism', () => {
    it('should produce identical scores for identical strategy (snapshot)', async () => {
      const mockStrategy: BrandStrategy = {
        purpose: 'Empower developers to build faster with AI-powered tools',
        mission: 'Democratize software development through intelligent automation',
        vision: 'A world where anyone can create professional software',
        values: ['Innovation', 'Accessibility', 'Quality', 'Community'],
        positioning: 'Premium developer tools for modern teams',
        personality: ['Innovative', 'Reliable', 'Professional', 'Friendly'],
        voiceAndTone: {
          voice: 'Expert but approachable',
          toneAttributes: ['Clear', 'Confident', 'Helpful']
        },
        keyMessages: [
          'Build faster with AI assistance',
          'Professional results, minimal effort',
          'Trusted by 10,000+ developers'
        ],
        differentiators: [
          'AI-powered code generation',
          'Seamless workflow integration',
          'Enterprise-grade security'
        ],
        proofPoints: [
          {
            claim: '10,000+ active developers',
            evidence: 'Internal analytics',
            confidence: 0.95
          },
          {
            claim: '40% faster development',
            evidence: 'Customer survey data',
            confidence: 0.85
          }
        ]
      };

      const result1 = await auditor.audit(mockStrategy, 'TestBrand', { mode: 'standard' });
      const result2 = await auditor.audit(mockStrategy, 'TestBrand', { mode: 'standard' });

      // Scores should be identical for same input
      expect(result1.overallScore).toBe(result2.overallScore);
      expect(result1.scoreBreakdown).toEqual(result2.scoreBreakdown);

      // Snapshot the complete scoring structure
      expect({
        overallScore: result1.overallScore,
        sourceQuality: result1.scoreBreakdown.sourceQuality.score,
        factVerification: result1.scoreBreakdown.factVerification.score,
        dataRecency: result1.scoreBreakdown.dataRecency.score,
        crossVerification: result1.scoreBreakdown.crossVerification.score,
        productionReadiness: result1.scoreBreakdown.productionReadiness.score,
      }).toMatchSnapshot();
    });

    it('should calculate weighted average correctly', async () => {
      const mockStrategy: BrandStrategy = {
        purpose: 'Test purpose',
        mission: 'Test mission',
        vision: 'Test vision',
      };

      const result = await auditor.audit(mockStrategy, 'TestBrand', { mode: 'standard' });

      // Verify weighted calculation matches expected formula
      // NOTE: The actual formula only uses 3 dimensions (dataRecency and crossVerification are hardcoded)
      // overallScore = sourceQuality * 0.3 + factVerification * 0.25 + productionReadiness * 0.45

      const expectedScore =
        result.scoreBreakdown.sourceQuality.score * 0.3 +
        result.scoreBreakdown.factVerification.score * 0.25 +
        result.scoreBreakdown.productionReadiness.score * 0.45;

      // Allow small floating point variance
      expect(Math.abs(result.overallScore - expectedScore)).toBeLessThan(0.1);
    });

    it('should assign correct status labels based on scores', async () => {
      const mockStrategy: BrandStrategy = {
        purpose: 'Test purpose',
      };

      const result = await auditor.audit(mockStrategy, 'TestBrand', { mode: 'standard' });

      // Verify status assignment logic matches implementation
      // Thresholds: >= 8 excellent, >= 6 good, >= 4 needs-work, < 4 critical
      Object.values(result.scoreBreakdown).forEach(dimension => {
        if (dimension.score >= 8) {
          expect(dimension.status).toBe('excellent');
        } else if (dimension.score >= 6) {
          expect(dimension.status).toBe('good');
        } else if (dimension.score >= 4) {
          expect(dimension.status).toBe('needs-work');
        } else {
          expect(dimension.status).toBe('critical');
        }
      });
    });

    it('should produce consistent findings structure', async () => {
      const mockStrategy: BrandStrategy = {
        purpose: 'Build the future of work',
        mission: 'Enable remote collaboration',
        values: ['Transparency', 'Innovation'],
      };

      const result = await auditor.audit(mockStrategy, 'TestBrand', { mode: 'comprehensive' });

      // Verify findings structure
      expect(result.findings).toBeInstanceOf(Array);
      result.findings.forEach(finding => {
        expect(finding).toHaveProperty('severity');
        expect(finding).toHaveProperty('category');
        expect(finding).toHaveProperty('message');
        expect(['critical', 'warning', 'info', 'success']).toContain(finding.severity);
        expect(['sources', 'facts', 'recency', 'verification', 'quality']).toContain(finding.category);
      });

      // Snapshot findings structure
      expect(result.findings.length).toMatchSnapshot();
    });

    it('should generate actionable recommendations', async () => {
      const mockStrategy: BrandStrategy = {
        purpose: 'Transform education',
        differentiators: ['AI tutoring', 'Personalized learning'],
      };

      const result = await auditor.audit(mockStrategy, 'TestBrand', { mode: 'standard' });

      // Verify recommendations structure
      expect(result.recommendations).toBeInstanceOf(Array);
      result.recommendations.forEach(rec => {
        expect(rec).toHaveProperty('priority');
        expect(rec).toHaveProperty('action');
        expect(rec).toHaveProperty('estimatedEffort');
        expect(rec).toHaveProperty('impact');
        expect(['high', 'medium', 'low']).toContain(rec.priority);
      });

      // Snapshot recommendation count
      expect(result.recommendations.length).toMatchSnapshot();
    });
  });

  describe('Score boundary conditions', () => {
    it('should handle minimal strategy (worst case)', async () => {
      const minimalStrategy: BrandStrategy = {};

      const result = await auditor.audit(minimalStrategy, 'MinimalBrand', { mode: 'quick' });

      expect(result.overallScore).toBeGreaterThanOrEqual(0);
      expect(result.overallScore).toBeLessThanOrEqual(10);
      expect(result.scoreBreakdown.productionReadiness.score).toBeLessThan(5);
    });

    it('should handle comprehensive strategy (best case)', async () => {
      const comprehensiveStrategy: BrandStrategy = {
        purpose: 'Clear purpose statement with evidence',
        mission: 'Specific mission with measurable outcomes',
        vision: 'Compelling vision for the future',
        values: ['Value1', 'Value2', 'Value3', 'Value4', 'Value5'],
        positioning: 'Unique positioning in competitive landscape',
        personality: ['Trait1', 'Trait2', 'Trait3'],
        voiceAndTone: {
          voice: 'Professional and approachable',
          toneAttributes: ['Clear', 'Confident', 'Inspiring']
        },
        keyMessages: [
          'Message 1 with data',
          'Message 2 with proof',
          'Message 3 with validation'
        ],
        differentiators: [
          'Unique differentiator 1',
          'Competitive advantage 2',
          'Market leadership 3'
        ],
        proofPoints: [
          { claim: 'Claim 1', evidence: 'Strong evidence', confidence: 0.95 },
          { claim: 'Claim 2', evidence: 'Research data', confidence: 0.90 },
          { claim: 'Claim 3', evidence: 'Third-party validation', confidence: 0.88 }
        ]
      };

      const result = await auditor.audit(comprehensiveStrategy, 'ComprehensiveBrand', { mode: 'comprehensive' });

      expect(result.overallScore).toBeGreaterThan(5);
      expect(result.scoreBreakdown.productionReadiness.score).toBeGreaterThan(7);
    });
  });

  describe('Mode-specific behavior', () => {
    it('should respect quick mode (faster, less thorough)', async () => {
      const strategy: BrandStrategy = { purpose: 'Test' };

      const result = await auditor.audit(strategy, 'TestBrand', { mode: 'quick' });

      expect(result.auditDate).toBeDefined();
      expect(result.brandName).toBe('TestBrand');
    });

    it('should respect comprehensive mode (thorough analysis)', async () => {
      const strategy: BrandStrategy = { purpose: 'Test' };

      const result = await auditor.audit(strategy, 'TestBrand', { mode: 'comprehensive' });

      expect(result.auditDate).toBeDefined();
      expect(result.recommendations.length).toBeGreaterThan(0);
    });
  });
});
