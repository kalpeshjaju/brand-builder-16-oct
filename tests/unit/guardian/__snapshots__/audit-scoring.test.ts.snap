// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Audit Scoring Arithmetic > Score calculation determinism > should generate actionable recommendations 1`] = `3`;

exports[`Audit Scoring Arithmetic > Score calculation determinism > should produce consistent findings structure 1`] = `2`;

exports[`Audit Scoring Arithmetic > Score calculation determinism > should produce identical scores for identical strategy (snapshot) 1`] = `
{
  "crossVerification": 6,
  "dataRecency": 7,
  "factVerification": 5,
  "overallScore": 6.4,
  "productionReadiness": 10,
  "sourceQuality": 2,
}
`;
